lib/storage/src/content_manager/consensus/consensus_wal.rs

```diff
@@ -1,203 +1,288 @@
+
+        self.entry_by_wal_index(wal_index)
     }
 
     pub fn entries(
         &self,
-        low: u64,
-        high: u64,
-        max_size: Option<u64>,
+        from_raft_index: u64,
+        until_raft_index: u64,
+        max_size_bytes: Option<u64>,
     ) -> raft::Result<Vec<RaftEntry>> {
-        let mut size = 0;
-
-        let entries = (low..high)
-            .map(|id| self.entry(id))
-            .take_while(|entry| {
-                let Ok(entry) = entry else { return true };
-
-                // It's somewhat unclear how `max_size == Some(0)` should be treated.
-                //
-                // `raft::storage::Storage::entries` documentation and `raft::util::limit_size` implementation
-                // suggest that it means "return only a single entry", but why not request a single entry
-                // using `low`/`high` arguments in that case? 🤔
-                //
-                // Another reasonable interpretation might be "unlimited" (e.g., same as `None` and `u64::MAX`).
-                //
-                // `raft::RawNode` would request ridiculous range (e.g., 1 mil entries) with `max_size == Some(0)`,
-                // which, IMO, does not make much sense. 😕
-                //
-                // The current implementation follows `raft::storage::Storage::entries` spec,
-                // but further research/experimentation might be needed.
-                //
-                // See:
-                // - https://docs.rs/raft/latest/raft/storage/trait.Storage.html#tymethod.entries
-                // - https://github.com/tikv/raft-rs/blob/v0.7.0/src/util.rs#L56-L71
-                let max_size = match max_size {
-                    Some(max_size) if max_size < u64::MAX => max_size,
-
-                    // It might also be reasonable to add a hard limit on the total size of returned entries here.
-                    //
-                    // For the same reason, that if `raft::RawNode` request some ginormous range without reasonable
-                    // `max_size` the current `entries` implementation will block consensus thread until the whole
-                    // range is collected.
-                    _ => return true,
-                };
-
-                let first_entry = size == 0;
-                size += u64::from(entry.compute_size());
-
-                size <= max_size || first_entry
-            })
-            .try_collect()?;
+        let offset = self.index_offset()?;
+
+        // Map requested Raft indices to WAL indices
+        let from_wal_index = offset.raft_to_wal(from_raft_index);
+        let until_wal_index = offset.raft_to_wal(until_raft_index);
+
+        // Bound mapped WAL indices between first/last WAL indices
+        let from_wal_index = cmp::max(from_wal_index, offset.wal_index);
+        let until_wal_index = cmp::min(until_wal_index, offset.wal_index + self.wal.num_entries());
```

Comment:
Is this the same? If so I think I'd find that a bit easier to read.

```suggestion
        let from_wal_index = cmp::max(from_wal_index, self.wal.first_index());
        let until_wal_index = cmp::min(until_wal_index, self.wal.last_index());
```

Reply:
It's not, _if we consider `until` to be exclusive_. `first_index + num_entries` always returns one-after-the-last-index.

Reply:
Yes, makes sense. Thanks!